src/git.rs

```diff
@@ -1,91 +1,145 @@
 use std::path::Path;
 
-use git2::build::RepoBuilder;
-use git2::{Error, Oid, Repository};
+use git2::build::{CheckoutBuilder, RepoBuilder};
+use git2::ResetType::Hard;
+use git2::{Error, Repository};
+use url::Url;
+
+pub struct Credentials {
+    pub login: String,
+    pub password: String,
+}
 
-/// TODO support SSH repository_url - we assume that the repository URL starts with HTTPS
-/// TODO support git submodules
 pub fn clone<P>(repository_url: &str, into_dir: P, credentials: &Option<Credentials>) -> Result<Repository, Error>
 where
     P: AsRef<Path>,
 {
-    let final_repository_url = match credentials {
-        Some(c) => format!(
-            "https://{}:{}@{}",
-            c.login,
-            c.password,
-            repository_url.replace("https://", "")
-        ),
-        None => repository_url.to_string(),
-    };
+    let mut url = Url::parse(repository_url)
+        .map_err(|err| Error::from_str(format!("Invalid repository url {}: {}", repository_url, err).as_str()))?;
 
-    RepoBuilder::new().clone(final_repository_url.as_str(), into_dir.as_ref())
-}
+    if url.scheme() != "https" {
```

Comment:
what if git?

Reply:
not supported,  the core never send a git@ url